### PR TITLE
Show implementations when URL derived state points to implementations

### LINKED ISSUE #44358: Other UI improvements - find implementations, display no implementations - must have - FE scale audit - strategic readiness
### Description
When there are no implementations found, the 'Find implementations' button in the tooltips leads to 'references' tab with 'references' and 'definitions' sections only. The 'implementations' section is not visible unless there are some implementations found.
- [loom](https://www.loom.com/share/d1059be95be4448ba8cbaeb4408c59aa) showing experience with no implementations section
- [example](https://sourcegraph.com/github.com/gorilla/mux@07eedffb4388b4ed26b86c67aedca1e513e7553b/-/blob/middleware.go?L14:6&subtree=true#tab=implementations_go) that has some implementations

Problem: this experience is confusing since when user clicks on 'Find implementations', they see no information about 0 implementations found. 

Proposed solution: Always show the 'implementations' tab. For cases with no implementations, display 0 and disable opening the section.

### Plan

- [ ] Update this issue with a design template
- [ ] Determine the milestone when it is delivered

---
Additional resources:
- [Product Experience at Scale: Great vs. Good vs. Bad](https://docs.google.com/document/d/1KPsPfZqZ2AAIJ5Ybe7QrWKuaafYF5vUcaB1IfiF82Po/edit#heading=h.4jfhf6ai2t49)
- [Raw scale audit](https://docs.google.com/spreadsheets/d/1tqpi22csQJNlk7Q9JFWcFj09VbWOiFdp-Vf18qJsWHc/edit#gid=0)

/cc @felixfbecker

### PR DESCRIPTION
**Pull request by philipp-spiess**
Fixes #44358

When a user clicks the "Go to implementations" link right now and the resulting references panel view does not contain any implementations, they land on a very confusing view that does not contain any information about the implementations feature:
 
Screenshot 2022-11-29 at 18 50 57

This PR fixes this issue by looking at the URL derived collapse state: If the URL describes a link to the implementations page (which is the case for this button) we always want to render the implementations view, even when there are no implementations displayed.

## Test plan
 
https://user-images.githubusercontent.com/458591/205055118-b86af2c6-22a2-44a2-ba71-46b9770e2307.mov

## App preview:

- [Web](https://sg-web-ps-fix-references-panel-go-to.onrender.com/search)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

File: client/web/src/codeintel/ReferencesPanel.tsx [hunk 1]
```diff
@@ -389,7 +389,7 @@ const ReferencesList: React.FunctionComponent<
     // Manual management of the open/closed state of collapsible lists so they
     // stay open/closed across re-renders and re-mounts.
     const location = useLocation()
-    const initialCollapseState = useMemo((): Record<string, boolean> => {
```
Comment (philipp-spiess) [inline client/web/src/codeintel/ReferencesPanel.tsx:392]: TS can infer the (exact) types anyways 🤷

**Comment by sg-e2e-regression-test-bob** [comment]
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.00% (0.00 kb) | 0.00% (+0.02 kb) | 0.00% (+0.02 kb) | 0.00% (0) |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/ps/fix-references-panel-go-to-implementations/compare-report.html;tab=live_object) for a full comparison between the commits 403976f and dcfb092 or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

**Review by taras-yemets — APPROVED** [review]
Nice! Thank you, @philipp-spiess! 
I have a concern about the "Implementations (0 items displayed)" label. Maybe we can be more specific here and say that no implementations were found?

**Comment by philipp-spiess** [comment]
@taras-yemets This is what happens:

Screenshot 2022-11-30 at 14 42 22

I added a note about the case where this does not work to the video above:

> Note: The first time I click Go to implementations in this video I do it on a symbol that already had the references panel open for it, in which case the auto-expand doesn't seem to work. This is an unrelated issue, you can see that it works for the second symbol.

**Review by AlicjaSuska — CHANGES_REQUESTED** [review]
Hi @philipp-spiess ! Thank you for making the changes. I think we should change the logic slightly and display the 'Implementations' section always when the bottom panel is open.

In the video that you've attached, the 'Implementations' section appears only after user clicks on 'Find implementations' but it doesn't appear when they click on 'Find references'. I think that it's not consistent - we should either display just one section that user 'requested' (not a good solution) or all sections every time we open the panel (I think this is the desired option).

Could you please make this change? Thank you!

**Comment by philipp-spiess** [comment]
Sure that's not a problem but I’m curious why we had this check in the first place, I'll do some research!

**Comment by philipp-spiess** [comment]
@AlicjaSuska Updated the PR and the test plan. Please re-review.

**Review by AlicjaSuska — APPROVED** [review]
Works great! Thank you @philipp-spiess

**Pull request merged**